hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java

```diff
@@ -385,32 +393,50 @@ public void deleteCache(String cacheNameWithPrefix, String callerUuid, boolean d
     public CacheConfig putCacheConfigIfAbsent(CacheConfig config) {
         // ensure all configs registered in CacheService are not PreJoinCacheConfig's
         CacheConfig cacheConfig = asCacheConfig(config);
-        CacheConfig localConfig = configs.putIfAbsent(cacheConfig.getNameWithPrefix(), cacheConfig);
-        if (localConfig == null) {
-            if (cacheConfig.isStatisticsEnabled()) {
-                setStatisticsEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
-            }
-            if (cacheConfig.isManagementEnabled()) {
-                setManagementEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+        CacheConfigFuture future = new CacheConfigFuture(nodeEngine, cacheConfig);
+        CacheConfigFuture localConfigFuture = configs.putIfAbsent(cacheConfig.getNameWithPrefix(), future);
+        // if the existing cache config future is not yet fully configured, we block here
+        CacheConfig localConfig = localConfigFuture == null ? null : localConfigFuture.join();
+        if (localConfigFuture == null) {
+            try {
+                if (cacheConfig.isStatisticsEnabled()) {
+                    setStatisticsEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+                }
+                if (cacheConfig.isManagementEnabled()) {
+                    setManagementEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+                }
+                logger.info("Added cache config: " + cacheConfig);
+                additionalCacheConfigSetup(null, config);
+                // now it is safe for others to obtain the new cache config
+                future.complete();
+            } catch (RuntimeException e) {
+                configs.remove(cacheConfig.getNameWithPrefix(), future);
+                future.complete(e);
+                throw e;
             }
-        }
-        if (localConfig == null) {
-            logger.info("Added cache config: " + cacheConfig);
+        } else {
+            additionalCacheConfigSetup(localConfig, config);
         }
         return localConfig;
     }
 
+    protected void additionalCacheConfigSetup(CacheConfig localConfig, CacheConfig config) {
```

Comment:
this may be clearer if we only pass `config` to this method?

Reply:
Yes, modified the signature to pass just the `config` object (be it already existing one or new one) and a boolean indicating whether this is a new or existing config.